src/demo.cc

```diff
@@ -0,0 +1,26 @@
+/*
```

Comment:
I think I'd structure the code a bit differently. How I imagined the structure of the c++ lib is something like this:

```
iceberg-cpp/
      - api/
          - src/
          - test/
...
      - core/
            - src/
            - test/
...
      - example/
            - src/
            - test/
```
and so on. (I hope the indentation is displayed as I intended to :) )

So in general I think there should be separate libs for each of the modules of the implementation and each of them should have their own src/ test/ folders. As I see now these folders are provided on the top level of the directory structure.
If we want to have an include/ folder for the headers separately, I think we should also have them on the module-level and not at the top level.

What do you think?

Reply:
Do we need more libraries other than `libiceberg` (which is the core library in your structure)? In my mind, `libiceberg` provides the building blocks for types, io, exceptions, metadata, transformations, expressions, table, catalog, reader/writer, etc.

Reply:
In my opinion we'd need at leas api/ and core/ (I'd prefer core/ over libiceberg just to be in line with the Java impl. No point of reinventing the wheel). Also I'd imagine puffin could also get a separate lib. And then if there is a need each engine could put their own connector into their own folders too.

Reply:
I'm not sure if I understand it correctly. Do you mean that the `api` folder stores all public headers? Then each folder (e.g. `core`, `puffin`, `example`, etc) uses its own subset of headers from the `api` folder to create library or executable?

Reply:
Personally I'm not in favor of splitting this repo into multiple libraries, which is an overkill. Originally I thought that the `api` and `core` folders from the apache/iceberg repo are all that we need. However, it seems that `puffin` is a good example which does not need to depend on other modules. Therefore I'm open to discuss.

Reply:
Ok ,what if we start with api/ core/ puffin/ and example/ and then we'll see if there is a need for anything else?

Reply:
Sounds good! It is still unclear what it will look like eventually. I'm adding third-party libraries (arrow, avro and simdjson and their indirect dependencies) now but let me finish this patch first.

Reply:
what is `api` in this context? is this the public file headers to include?

Reply:
What about this?
```
iceberg-cpp/
├── api/
│   ├── table.h
│   └── puffin.h
├── example/
│   └── demo.cc
├── src/
│   ├── core/
│   │   ├── base_table.h
│   │   └── base_table.cc
│   └── puffin/
│       └── puffin.cc
└── test/
```

Reply:
Seems great!